### PR TITLE
Add some basic preferences for the monaco editor

### PR DESCRIPTION
**Pull request by epatpol**
This adds the possibility of changing the tabSize and showing/hiding the
lineNumbers by modifying the .theia/settings.json file.

Signed-off-by: Patrick-Jeffrey Pollo Guilbert <patrick.pollo.guilbert@ericsson.com>

### CODE REVIEW

**Comment by epatpol** [comment]
This is based on @akosyakov's work for the filesystemwatcher. It's still a WIP, so don't merge yet, but feel free to add comments.

**Comment by akosyakov** [comment]
so it should be:
```ts
export interface EditorConfiguration {
    'editor.tabSize': number,
    'editor.lineNumbers': 'on' | 'off'
}
```

**Comment by epatpol** [comment]
That makes sense, that's what I thought after pushing the patch also (putting more generic preferences in editor and more specific preferences in the monaco/X editor implementation), I'll change it asap.

File: src/monaco/browser/monaco-editor-provider.ts [hunk 1]
```diff
@@ -40,16 +43,27 @@ export class MonacoEditorProvider {
                     wordWrap: false,
                     folding: true,
                     theme: 'vs-dark',
-                    readOnly: model.readOnly
+                    readOnly: model.readOnly,
```
Comment (akosyakov) [inline src/monaco/browser/monaco-editor-provider.ts]: don't forget to properly initialize the editor initially

File: src/monaco/browser/monaco-editor-provider.ts [hunk 2]
```diff
@@ -40,16 +43,27 @@ export class MonacoEditorProvider {
...
                     textModelResolverService: this.monacoModelResolver,
                     contextMenuService: this.contextMenuService,
                     commandService
                 }
             );
+
+
+            this.monacoPrefs.onPreferenceChanged(e => {
+                if (e.preferenceName === 'monaco.tabSize') {
+                    editor.getControl().getModel().updateOptions({ tabSize: <number>e.newValue });
```
Comment (akosyakov) [inline src/monaco/browser/monaco-editor-provider.ts]: we should try to find a way to get rid of type casting, but separately

**Comment by epatpol** [comment]
Changed it to be more generic, and added the whitespaces option (thx @akosyakov)

working example:


screencast_2017-07-10_04-29-52 pm

File: src/monaco/browser/monaco-editor-provider.ts [hunk 3]
```diff
@@ -25,7 +27,8 @@ export class MonacoEditorProvider {
         @inject(MonacoToProtocolConverter) protected readonly m2p: MonacoToProtocolConverter,
         @inject(ProtocolToMonacoConverter) protected readonly p2m: ProtocolToMonacoConverter,
         @inject(MonacoWorkspace) protected readonly workspace: MonacoWorkspace,
-        @inject(MonacoCommandServiceFactory) protected readonly commandServiceFactory: MonacoCommandServiceFactory
+        @inject(MonacoCommandServiceFactory) protected readonly commandServiceFactory: MonacoCommandServiceFactory,
+        @inject(EditorPreferences) protected readonly monacoPrefs: EditorPreferences,
```
Comment (akosyakov) [inline src/monaco/browser/monaco-editor-provider.ts]: `editorPreferneces` or `preferences`

File: src/editor/browser/index.ts [hunk 4]
```diff
@@ -13,3 +13,4 @@ export * from './editor-command';
 export * from './editor-menu';
 export * from './editor-keybinding';
 export * from './editor-frontend-module';
+export * from './editor-preferences';
```
Comment (akosyakov) [inline src/editor/browser/index.ts]: Where `editor-preferences` itself? I don't see it in the commit

File: src/monaco/browser/monaco-editor-provider.ts [hunk 5]
```diff
@@ -48,13 +51,33 @@ export class MonacoEditorProvider {
                     commandService
                 }
             );
-            editor.onDispose(() => reference.dispose());
 
+            this.monacoPrefs.onPreferenceChanged(e => {
```
Comment (akosyakov) [inline src/monaco/browser/monaco-editor-provider.ts]: you should dispose a listener if the editor is disposed

Comment (akosyakov) [inline src/monaco/browser/monaco-editor-provider.ts]: It is not enough listening to changes, the editor should be initialized initially properly.

File: src/monaco/browser/monaco-frontend-module.ts [hunk 6]
```diff
@@ -20,13 +20,16 @@ import { MonacoEditorService } from "./monaco-editor-service";
 import { MonacoModelResolver } from "./monaco-model-resolver";
 import { MonacoContextMenuService } from "./monaco-context-menu";
 import { MonacoCommandService, MonacoCommandServiceFactory } from './monaco-command-service';
+import { bindEditorPreferences } from '../../editor/browser/editor-preferences'
```
Comment (akosyakov) [inline src/monaco/browser/monaco-frontend-module.ts]: it should be bound in the editor extension, the monaco extension assume that it is bound.

File: src/monaco/browser/monaco-editor-provider.ts (hunk 5 above)
Comment (epatpol) [inline src/monaco/browser/monaco-editor-provider.ts]: Oh right true that, right now the prefs would only load on settings.change

**Comment by akosyakov** [comment]
only wonder what happens if the bogus value is entered

**Comment by epatpol** [comment]
Hmmm true that, it should fall back to the default value but right now nothing happens.

**Comment by akosyakov** [comment]
@epatpol just file a bug

**Comment by epatpol** [comment]
Alright then I'll merge this one and open a bug thanks @akosyakov

**Pull request merged**